engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateByWindowTime.java

```diff
@@ -0,0 +1,400 @@
+            }
+        }
+
+        // naturally need to compute all newly added rows
+        tmpAffected.insert(upstream.added());
+
+        ctx.affectedRows = tmpAffected;
+
+        // now get influencer rows for the affected rows
+        ctx.influencerRows = computeInfluencerRowsTime(ctx.sourceRowSet, ctx.affectedRows, prevUnits, fwdUnits,
+                ctx.timestampColumnSource, ctx.timestampSsa, false);
+
+        makeOperatorContexts(ctx);
+    }
+
+    private long advanceToNonNull(LongColumnIterator it) {
+        long value;
+        do {
+            value = it.hasNext() ? it.nextLong() : Long.MAX_VALUE;
+        } while (value == NULL_LONG);
+        return value;
+    }
+
+    /***
+     * This function process the affected rows chunkwise, and will advance the moving window (which is the same for all
+     * operators in this collection). For each row in the dataset the sliding window will adjust and instructions for
+     * pushing/popping data will be created for the operators. For each chunk of `affected` rows, we will identify
+     * exactly which `influencer` rows are needed and will provide those and the push/pop instructions to the operators.
+     *
+     * Downstream operators should manage local storage in a RingBuffer or other efficient structure since our pop()
+     * calls do not provide the popped data
+     */
+    @Override
+    public void processRows(UpdateByWindowBucketContext context, boolean initialStep) {
+        UpdateByWindowBucketTimeContext ctx = (UpdateByWindowBucketTimeContext) context;
+
+        for (int opIdx : context.dirtyOperatorIndices) {
+            UpdateByWindowedOperator winOp = (UpdateByWindowedOperator) operators[opIdx];
+            // call the specialized version of `intializeUpdate()` for these operators
+            winOp.initializeUpdate(ctx.opContext[opIdx]);
+        }
+
+        try (final RowSequence.Iterator it = ctx.affectedRows.getRowSequenceIterator();
+                final LongColumnIterator influencerTsHeadIt = new LongColumnIterator(context.timestampColumnSource, context.influencerRows);
+                final LongColumnIterator influencerTsTailIt = new LongColumnIterator(context.timestampColumnSource, context.influencerRows);
+                final RowSequence.Iterator influencerKeyIt = ctx.influencerRows.getRowSequenceIterator();
+                final ChunkSource.GetContext localTimestampContext =
+                        ctx.timestampColumnSource.makeGetContext(ctx.workingChunkSize);
+                final WritableIntChunk<? extends Values> pushChunk =
+                        WritableIntChunk.makeWritableChunk(ctx.workingChunkSize);
+                final WritableIntChunk<? extends Values> popChunk =
+                        WritableIntChunk.makeWritableChunk(ctx.workingChunkSize)) {
+
+            long currentHeadTs = advanceToNonNull(influencerTsHeadIt);
+            long currentTailTs = advanceToNonNull(influencerTsTailIt);
```

Comment:
I wonder if we could come up with a strategy to clone the head iterator at this stage.

Reply:
OBE, using `LongRingBuffer` to store ts window data